### PR TITLE
Minimal test cleanup for g3.

### PR DESCRIPTION
**Pull request by hugelgupf**

### CODE REVIEW

File: cmds/which/which.go [hunk 1]
```diff
@@ -65,5 +64,8 @@ func main() {
 		p = "/ubin"
 	}
 
-	which(p, os.Stdout, flag.Args())
+	if err := which(os.Stdout, strings.Split(p, ":"), flag.Args(), allPaths); err != nil {
```
Comment (rminnich) [inline cmds/which/which.go]: Will this break things?
current behavior
rminnich@uroot:~/go/src/github.com/u-root/elvish$ which adfa
rminnich@uroot:~/go/src/github.com/u-root/elvish$ 

no error printed.

Comment (rminnich) [reply cmds/which/which.go]: oh thank you for formatting that for me github.

 `rminnich@uroot:~/go/src/github.com/u-root/elvish$ which adfa
rminnich@uroot:~/go/src/github.com/u-root/elvish$ `

Comment (hugelgupf) [reply cmds/which/which.go]: On my machine:

```
hugelgupf :: which foo
foo not found                                                                                                                                                                                                                                
hugelgupf ::  
```

Comment (rminnich) [reply cmds/which/which.go]: rminnich@uroot:~/go/src/github.com/u-root/u-root$ /usr/bin/which foo
rminnich@uroot:~/go/src/github.com/u-root/u-root$

**Comment by rminnich** [comment]
did you get a chance to test /usr/bin/which?

**Comment by rminnich** [comment]
so, um, I think it's clear that /usr/bin/which returns any error in the exit status, but prints nothing. can we wrap this PR up with that behavior?

File: cmds/which/which.go (hunk 1 above)
Comment (hugelgupf) [reply cmds/which/which.go]: yea that was zsh.

**Comment by hugelgupf** [comment]
ptal

**Comment by hugelgupf** [comment]
This is now the minimal test cleanup needed to import u-root stuff into google3. I'd like to clean up more of this stuff, but this is the bare minimum necessary.

File: cmds/which/which.go [hunk 2]
```diff
@@ -21,38 +22,33 @@ import (
 )
 
 var (
-	flags struct {
-		allPaths bool
-	}
+	allPaths = flag.Bool("a", false, "print all matching pathnames of each argument")
+	verbose  = flag.Bool("v", false, "verbose output")
```
Comment (rminnich) [inline cmds/which/which.go]: This is not a standard switch but since you really seem to want this :-)

File: cmds/which/which.go [hunk 3]
```diff
@@ -21,38 +22,33 @@ import (
...
-}
-
-func which(p string, writer io.Writer, cmds []string) {
-	pathArray := strings.Split(p, ":")
-
-	// If no matches are found will exit 1, else 0
-	exitValue := 1
+func which(writer io.Writer, paths []string, cmds []string, allPaths bool) error {
+	var foundOne bool
 	for _, name := range cmds {
```
Comment (rminnich) [inline cmds/which/which.go]: and we're not using LookPath because we pass the paths in?

File: cmds/which/which_linux.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Execute permission bits.
+const (
+	otherExec = 1
+	groupExec = 1 << 3
+	userExec  = 1 << 6
+)
+
+func canExecute(path string) bool {
```
Comment (rminnich) [inline cmds/which/which_linux.go]: This should just use faccess at, e.g.
faccessat(AT_FDCWD, "/bin/date", X_OK)  = 0

File: cmds/which/which_other.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// +build !linux
+// Copyright 2016-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+func canExecute(path string) bool {
+	return true
```
Comment (rminnich) [inline cmds/which/which_other.go]: actually the first order approx in the linux one would be much better here.

File: cmds/which/which_linux.go (hunk 4 above)
Comment (hugelgupf) [reply cmds/which/which_linux.go]: faccessat does not to what you want it to do. It won't tell if you can actually exec the file. For one, faccessat uses the real UID/GID, while execve will use the effective UID/GID. the access family of syscalls is mostly useful for setuid programs.

File: cmds/which/which_other.go (hunk 5 above)
Comment (hugelgupf) [reply cmds/which/which_other.go]: I'll make _linux into _unix.

File: cmds/which/which.go (hunk 3 above)
Comment (hugelgupf) [reply cmds/which/which.go]: I didn't know LookPath was a thing! that's pretty neat. Yea, we also have a different exec approximation.

File: cmds/which/which_other.go (hunk 5 above)
Comment (rminnich) [reply cmds/which/which_other.go]: The problem with lookpath is that its test for executability is wrong (I tested it) and we'd be better off with NOT using it :-(
maybe clone it ... I would say let's not change which for now, but fix it better later with fxstatat?

**Pull request merged**